app/src/main/java/woowacourse/movie/presenter/MovieMainPresenter.kt

```diff
@@ -0,0 +1,16 @@
+package woowacourse.movie.presenter
+
+import android.content.Context
+import woowacourse.movie.model.MovieAdapter
+import woowacourse.movie.model.MovieRepository
+
+class MovieMainPresenter(private val movieChoiceContractView: MovieMainContract.View) :
+    MovieMainContract.Presenter {
+    private val movieRepository: MovieRepository = MovieRepository()
+    private lateinit var movieAdapter: MovieAdapter
+
+    fun getAdapter(context: Context): MovieAdapter {
+        movieAdapter = MovieAdapter(context, movieChoiceContractView, movieRepository.getAll())
```

Comment:
context 를 항상 꼭 넘겨야 할까요? 
모든 view 는 context 를 가지고 있다는 점 생각해보시면 좋을 것 같습니다.

Reply:
이 부분에 고민이 있었는데요 ㅜㅜ

MVP 패턴은 Presenter가 View와 Model 을 모두 알고 있는(=참조하는) 방식이라고 이해했습니다.
"그럼 Presenter가 Activity의 context를 알고, Activity의 생명주기를 결국에 따라야 한다는 거네?" 라는 생각이 들더라구요.

근데 어째서인지 Presenter를 생성할 때 Activity의 context 자체를 넘겨주는 행위는 뭔가 권장되지 않는 것 같고.. 지금과 같이 
명시적인 액티비티가 아닌, MovieMainContract.View 등으로 추상화해서 넘겨주면서 '이건 이 역할을 하는 인터페이스의 구현체야' 라고 Activity를 넘겨주는게 더 자연스럽다고 느꼈습니다.
[참고1](https://stackoverflow.com/questions/39263077/is-it-a-good-practice-to-access-the-actvitys-context-in-the-presenter-class-if)

같은 맥락으로 보니, MovieMainContract.View 를 구현한 구현체가 항상 컴포넌트라면 context를 내부적으로 갖고 있겠지만
그렇지 않을 경우에는 context가 없을 수도 있겠다는 생각을 해서, 따로 또 context를 받아보게 되었습니다.
제가 너무 일어나지 않을 일을 추상적으로 생각하고 있는 것일까요?

Reply:
> MVP 패턴은 Presenter가 View와 Model 을 모두 알고 있는(=참조하는) 방식이라고 이해했습니다.
"그럼 Presenter가 Activity의 context를 알고, Activity의 생명주기를 결국에 따라야 한다는 거네?" 라는 생각이 들더라구요.

그렇다면 팡태가 생각하시기에 Presenter 와 Controller 의 차이가 무엇인가요?
Controller 또한 View 와 Model 을 모두 알고 참조합니다.

MVP, MVVM, MVI 등의 아키텍처 모두 Model 과 View 를 분리하려는 MVC 에서 시작하였습니다.
하지만 MVC 에서 해결하지 못한 문제를 개선하기 위해,
추후의 아키텍처들이 나오게 된 것입니다.

가령 MVP 의 경우는 Controller 가 Activity 와 너무 밀접한 관련이 있다는 문제를 해결해보려고 나온 디자인패턴입니다.

<img width="806" alt="스크린샷 2024-04-20 오전 10 39 32" src="https://github.com/ghojeong/woowacourse-android/assets/38595650/929d27b5-78bd-4631-b897-c71f25a6319f">

그렇다면 MVP 는 이를 어떻게 해결한 것일까요?
그리고 이를 해결하기 위해 Presenter 와 Contract 에 어떤 역할을 부여하고, 역할 분리를 한 것일까요?

> MovieMainContract.View 를 구현한 구현체가 항상 컴포넌트라면 context를 내부적으로 갖고 있겠지만
그렇지 않을 경우에는 context가 없을 수도 있겠다는 생각을 해서, 따로 또 context를 받아보게 되었습니다.
제가 너무 일어나지 않을 일을 추상적으로 생각하고 있는 것일까요?

context 가 없을 수도 있는 경우는 물론 있습니다.
하지만 context 가 없다고 문제가 생기는게, 이미 심각한 의존성이 있다는 생각이 듭니다.
MVP 는 MVC 와 다르게 해당 부분에 대한 문제를 이미 해결한 아키텍처입니다.
그렇기에 Presenter 가 Activity 의 context 에 의존성이 생기는 일은 권장되지 않습니다.
["You Presenter (base or otherwise) should not know about Context, Activity, Toast or anything else Android based."](https://stackoverflow.com/questions/39263077/is-it-a-good-practice-to-access-the-actvitys-context-in-the-presenter-class-if)

MVC 가 있음에도 MVP 가 탄생된 배경과,
MVP 가 해결한 문제에 대해 더 깊게 공부해보시면 좋을 것 같습니다.
가능하시면 공부한 내용에 대한 레퍼런스도 공유해주시면 좋을 것 같습니다.